hadoop-hdds/docs/content/design/leader-execution/leader-execution.md

```diff
@@ -0,0 +1,422 @@
+}
+```
+
+Refer [Execution Control Message](leader-execution.md#control-request) for format.
+
+DB Persistence:
+```
+New rocksDB table: ClientRequestReplyTable
+Key: String: ClientId#CallId
+Value: ClientRequestInfo
+```
+
+#### Memory caching:
+```
+Memory Map: ClientId#CallId Vs Response
+Expiry: 10 minute (as current default for ratis)
+```
+
+Memory expiry handling:
+```
+Handling cleanup of expired entry:
+Timer: 30 sec
+RequestReply Bucket Map:
+Time vs List<ClientId#CallId>, where Time is with granularity of 10 sec
+```
+
+Steps:
+- Retrieve all bucket for time < (currTime - 10min as expiry)
+- Delete from DB and Map for the entry
+
+### Testability framework
+
+With rework on flow, a testability framework for better test coverage for request processing.
+
+Complexity in existing framework for request:
+1. Flow handling is different 1 node / 3 node HA deployment
+2. Check for double buffer cache flush
+3. cache related behaviour testing
+3. Ratis sync and failure handling
+4. Too much mocking for unit testing
+
+Proposed handling:
+Since execution is leader side and only db update is synchronized to all other nodes, so unit test will focus behavior, but not on env.
+
+1. Test data preparation
+  - Utility to prepare Key, File, volume, bucket
+  - Insert to DB to different table
+2. Short-circuit for db update (without ratis)
+3. simplified mocking for ranger authorization (just mock a method)
+4. behavior test from End-to-End perspective
+
+This will have the following advantages:
+- Speed of test scenarios (sync wait and double buffer flush wait will be avoided)
+- optimized test cases (avoid duplicate test cases)
+- test code complexity will be less
+
+TODO: With echo, define test utils and sample test cases
+- Capture test cases based on behavior.
+
+## Step-by-step integration of existing request (interoperability)
+
+Leader side execution have changes for all flows. This needs to be done incrementally to help better quality, testability.
+This needs below integration points in current code:
+1. dependency over ratis index removal (for old flow also)
+2. bucket quota handling integration (such that old flow have no impact)
+3. Granular locking for old flow, this ensures `no cache` for new flow do not have impact
+4. OmStateMachine integration for new flow, so that both old and new flow can work together
+5. Request segregation for new flow which is incrementally added.
+
+With above, Enable for old and new flow execution will be done with Feature flag, to switch between them seamlessly.
+And old flow can be removed with achieving quality, performance and compatibility for new flow execution.
+
+## Impacted areas
+1. With Leader side execution, metrics and its capturing information can change.
+   - Certain metrics may not be valid
+   - New metrics needs to be added
+   - Metrics will be updated at leader side now like for key create. At follower node, its just db update, so value will not be udpated.
```

Comment:
FYI, also note that write audit logs will only be generated in leader, instead of both in leader and follower.

Reply:
updated